src/acp/rule.test.ts

```diff
@@ -1351,6 +1355,146 @@ describe("getClientAll", () => {
+  });
+
+  it("does not change the input rule", () => {
+    const rule = mockRule(MOCKED_RULE_IRI, {
+      clients: [MOCK_CLIENT_WEBID_1],
+    });
+    setClient(rule, MOCK_CLIENT_WEBID_2.value);
+    expect(
+      rule.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_2)
+      )
+    ).toBe(false);
+    expect(
+      rule.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_1)
+      )
+    ).toBe(true);
+  });
+
+  it("does not overwrite the public client class", () => {
+    const rule = mockRule(MOCKED_RULE_IRI, {
+      publicClient: true,
+    });
+    const result = setClient(rule, MOCK_CLIENT_WEBID_1.value);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, SOLID_PUBLIC_CLIENT)
+      )
+    ).toBe(true);
+  });
+});
+
+describe("addClient", () => {
+  it("adds the given client to the rule", () => {
+    const rule = mockRule(MOCKED_RULE_IRI);
+    const result = addClient(rule, MOCK_CLIENT_WEBID_1.value);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_1)
+      )
+    ).toBe(true);
+  });
+
+  it("does not override existing clients/the public client class", () => {
+    const rule = mockRule(MOCKED_RULE_IRI, {
+      clients: [MOCK_CLIENT_WEBID_1],
+      publicClient: true,
+    });
+    const result = addClient(rule, MOCK_CLIENT_WEBID_2.value);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_1)
+      )
+    ).toBe(true);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_2)
+      )
+    ).toBe(true);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, SOLID_PUBLIC_CLIENT)
+      )
+    ).toBe(true);
+  });
+});
+
+describe("removeClient", () => {
+  it("removes the given client from the rule", () => {
+    const rule = mockRule(MOCKED_RULE_IRI, {
+      clients: [MOCK_CLIENT_WEBID_1],
+    });
+    const result = removeClient(rule, MOCK_CLIENT_WEBID_1.value);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_1)
+      )
+    ).toBe(false);
+  });
+
+  it("does not delete unrelated clients", () => {
+    const rule = mockRule(MOCKED_RULE_IRI, {
+      clients: [MOCK_CLIENT_WEBID_1, MOCK_CLIENT_WEBID_2],
+      publicClient: true,
+    });
+    const result = removeClient(rule, MOCK_CLIENT_WEBID_2.value);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_2)
+      )
+    ).toBe(false);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, MOCK_CLIENT_WEBID_1)
+      )
+    ).toBe(true);
+    expect(
+      result.has(
+        DataFactory.quad(MOCKED_RULE_IRI, ACP_CLIENT, SOLID_PUBLIC_CLIENT)
+      )
+    ).toBe(true);
+  });
+
+  it("does not remove groups, even with a matching IRI", () => {
```

Comment:
This could also technically check if it doesn't remove agents, but that's being overly precautious so I wouldn't mind if it doesn't.

Reply:
Fair enough, that was lazy copy-pasting and I shouldn't be lazy: 2e246497